x-pack/plugins/security_solution/server/endpoint/services/actions/action_list.ts

```diff
@@ -0,0 +1,179 @@
+  logger,
+  size,
+  startDate,
+  userIds,
+}: GetActionDetailsListParam): Promise<ActionDetails[]> => {
+  let actions;
+  let actionReqIds;
+
+  try {
+    // fetch actions with matching agent_id
+    const { actionIds, actionRequests } = await getActionsForAgents({
+      actionTypes,
+      context,
+      logger,
+      elasticAgentIds,
+      startDate,
+      endDate,
+      from,
+      size,
+      userIds,
+    });
+    actions = actionRequests;
+    actionReqIds = actionIds;
+  } catch (error) {
+    logger.error(error);
+    throw error;
+  }
+
+  if (actions?.statusCode !== 200) {
+    logger.error(`Error fetching actions log for agent_ids ${elasticAgentIds}`);
+    throw new Error(`Error fetching actions log for agent_ids ${elasticAgentIds}`);
+  }
+
+  // categorize actions as fleet and endpoint actions
+  const categorizedActions = categorizeActionResults({
+    results: actions?.body?.hits?.hits,
+  });
+
+  // normalized actions with a flat structure to access relevant values
+  const normalizedActionRequests: Array<ReturnType<typeof mapToNormalizedActionRequest>> =
+    categorizedActions.map((action) => mapToNormalizedActionRequest(action.item.data));
+
+  // get all responses for given action Ids and agent Ids
+  const actionResponses = await getActionResponses({
+    context,
+    logger,
+    elasticAgentIds,
+    actionIds: actionReqIds,
+  });
+
+  // categorize responses as fleet and endpoint responses
+  const categorizedResponses = categorizeResponseResults({
+    results: actionResponses?.body?.hits?.hits,
+  });
+
+  // compute action details list for each action id
+  const actionDetails: ActionDetails[] = normalizedActionRequests.map((action) => {
+    // pick only those actions that match the current action id
+    const matchedActions = categorizedActions.filter((ac) =>
+      ac.type === 'action'
+        ? ac.item.data.EndpointActions.action_id === action.id
+        : ac.item.data.action_id === action.id
+    );
+    // pick only those responses that match the current action id
+    const matchedResponses = categorizedResponses.filter((re) =>
+      re.type === 'response'
+        ? re.item.data.EndpointActions.action_id === action.id
+        : re.item.data.action_id === action.id
+    );
+
+    // find the specific response's details using that set of matching responses
+    const { isCompleted, completedAt, wasSuccessful, errors } = getActionCompletionInfo(
+      action.agents,
+      matchedResponses
+    );
+
+    return {
+      id: action.id,
+      agents: action.agents,
+      command: action.command,
+      startedAt: action.createdAt,
+      logEntries: [...matchedActions, ...matchedResponses],
+      isCompleted,
+      completedAt,
+      wasSuccessful,
+      errors,
+      isExpired: !isCompleted && action.expiration < new Date().toISOString(),
+    };
+  });
+
+  // sort the list by startedAt in desc order, newest first
+  const sortedData = getTimeSortedActionDetails(actionDetails);
```

Comment:
this should not be necessary. The elasticsearch query (for the action requests) should sort the results

Reply:
Good catch. I should be sorting the log entries actually. Will change this.